### PR TITLE
HDDS-8484. Allow Ozone volume name to have special characters

### PR DESCRIPTION
**Pull request by DaveTeng0**
## What changes were proposed in this pull request?
Allow ozone user to create volume name which contains non-s3-naming-convention character (e.g. underscore)

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-8484

## How was this patch tested?
unit tests

### CODE REVIEW

**Comment by DaveTeng0** [comment]
(cc. @umamaheswararao @jojochuang @arp7 @kerneltime @adoroszlai

**Review by adoroszlai — CHANGES_REQUESTED** [review]
Thanks @DaveTeng0 for the patch.  I've noticed two issues:

* Some `RpcClient` methods call `verifyResourceName` instead of `verifyVolumeName`.  These methods still reject non-strict names, since those calls were not removed.
* nit: The method `verifyVolumeName` is no longer used, can be removed.

On second thought, most of the name checks still apply to volume or bucket names regardless of the strictness setting.  E.g. name should not be `null`, too short, too long, etc.

Removing client-side validation introduces two problems:

1. usability: previously these were rejected at the client with specific message, but now we only get a generic message from OM
2. performance: validation requires round-trip to OM

Before:

```
$ ozone sh volume create /v
INVALID_VOLUME_NAME Bucket or Volume length is illegal, valid length is 3-63 characters
$ ozone sh volume create /.hidden
INVALID_VOLUME_NAME Bucket or Volume name cannot start with a period or dash
$ ozone sh volume create /toooooooooooooooooooooooooooooooooooooooooooooooooolooooooooooooooooooooooooooooooooooooooooooooong
INVALID_VOLUME_NAME Bucket or Volume length is illegal, valid length is 3-63 characters
$ ozone sh volume create 192.168.1.1
INVALID_VOLUME_NAME Bucket or Volume name cannot be an IPv4 address or all numeric
$ ozone sh volume create 12345
INVALID_VOLUME_NAME Bucket or Volume name cannot be an IPv4 address or all numeric
```

After:

```
$ ozone sh volume create /v
INVALID_VOLUME_NAME Invalid volume name: v
$ ozone sh volume create /.hidden
INVALID_VOLUME_NAME Invalid volume name: .hidden
$ ozone sh volume create /toooooooooooooooooooooooooooooooooooooooooooooooooolooooooooooooooooooooooooooooooooooooooooooooong
INVALID_VOLUME_NAME Invalid volume name: toooooooooooooooooooooooooooooooooooooooooooooooooolooooooooooooooooooooooooooooooooooooooooooooong
$ ozone sh volume create 192.168.1.1
INVALID_VOLUME_NAME Invalid volume name: 192.168.1.1
$ ozone sh volume create 12345
INVALID_VOLUME_NAME Invalid volume name: 12345
```

https://github.com/apache/ozone/blob/3858cd102bbc3c328c24261d929c6549f5358a40/hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java#L491-L496

(The same applies to bucket name validation removed in HDDS-7586.)

So I think we should restore client-side validation (both volume and bucket), and always perform it with `strict=false`.  This catches most cases at the client with the more specific message.  OM will still reject names with `_` if it's set to `strict=true`.

**Comment by umamaheswararao** [comment]
>So I think we should restore client-side validation (both volume and bucket), and always perform it with strict=false

@adoroszlai I understand your concern. You mean we want to validate when strict = true? I think strict is true is old behavior right. IIUC, One problem here is, this flag is actually targeted at server side, so it's not necessarily available at client right?

**Comment by adoroszlai** [comment]
> You mean we want to validate when strict = true?

No, I mean that most of the name restrictions still apply regardless of the `strict` setting.  And now they are validated only at server-side and rejected with less useful message.

**Comment by DaveTeng0** [comment]
> Thanks @DaveTeng0 for the patch. I've noticed two issues:
> 
> * Some `RpcClient` methods call `verifyResourceName` instead of `verifyVolumeName`.  These methods still reject non-strict names, since those calls were not removed.
> * nit: The method `verifyVolumeName` is no longer used, can be removed.
> 
> On second thought, most of the name checks still apply to volume or bucket names regardless of the strictness setting. E.g. name should not be `null`, too short, too long, etc.
> 
> Removing client-side validation introduces two problems:
> 
> 1. usability: previously these were rejected at the client with specific message, but now we only get a generic message from OM
> 2. performance: validation requires round-trip to OM
> 
> Before:
> 
> ```
> $ ozone sh volume create /v
> INVALID_VOLUME_NAME Bucket or Volume length is illegal, valid length is 3-63 characters
> $ ozone sh volume create /.hidden
> INVALID_VOLUME_NAME Bucket or Volume name cannot start with a period or dash
> $ ozone sh volume create /toooooooooooooooooooooooooooooooooooooooooooooooooolooooooooooooooooooooooooooooooooooooooooooooong
> INVALID_VOLUME_NAME Bucket or Volume length is illegal, valid length is 3-63 characters
> $ ozone sh volume create 192.168.1.1
> INVALID_VOLUME_NAME Bucket or Volume name cannot be an IPv4 address or all numeric
> $ ozone sh volume create 12345
> INVALID_VOLUME_NAME Bucket or Volume name cannot be an IPv4 address or all numeric
> ```
> 
> After:
> 
> ```
> $ ozone sh volume create /v
> INVALID_VOLUME_NAME Invalid volume name: v
> $ ozone sh volume create /.hidden
> INVALID_VOLUME_NAME Invalid volume name: .hidden
> $ ozone sh volume create /toooooooooooooooooooooooooooooooooooooooooooooooooolooooooooooooooooooooooooooooooooooooooooooooong
> INVALID_VOLUME_NAME Invalid volume name: toooooooooooooooooooooooooooooooooooooooooooooooooolooooooooooooooooooooooooooooooooooooooooooooong
> $ ozone sh volume create 192.168.1.1
> INVALID_VOLUME_NAME Invalid volume name: 192.168.1.1
> $ ozone sh volume create 12345
> INVALID_VOLUME_NAME Invalid volume name: 12345
> ```
> 
> https://github.com/apache/ozone/blob/3858cd102bbc3c328c24261d929c6549f5358a40/hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OmUtils.java#L491-L496
> 
> (The same applies to bucket name validation removed in [HDDS-7586](https://issues.apache.org/jira/browse/HDDS-7586).)
> 
> So I think we should restore client-side validation (both volume and bucket), and always perform it with `strict=false`. This catches most cases at the client with the more specific message. OM will still reject names with `_` if it's set to `strict=true`.

good catch!

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/volume/TestOMVolumeCreateRequest.java [hunk 1]
```diff
@@ -210,6 +210,78 @@ public void testValidateAndUpdateCacheWithVolumeAlreadyExists()
         omMetadataManager.getVolumeKey(volumeName)));
   }
 
+  @Test
```
Comment (devmadhuu) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/volume/TestOMVolumeCreateRequest.java:213]: @DaveTeng0 - can we add test case for OMTenantCreateRequest as well ?

**Review by ashishkumar50 — COMMENTED** [review]
Hi @DaveTeng0 , Thanks for the patch. Can you add this check in TestMultiTenantVolume as well which is meant for S3 multi tenant volume scenario.

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java [hunk 2]
```diff
@@ -619,7 +620,6 @@ public void createBucket(String volumeName, String bucketName)
   public void createBucket(
       String volumeName, String bucketName, BucketArgs bucketArgs)
       throws IOException {
-    verifyVolumeName(volumeName);
```
[the remark this answers is not in the document]
Comment (DaveTeng0) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java:622]: oops unintentionally. Thanks attila!

[the remark this answers is not in the document]
Comment (adoroszlai) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java:622]: Let's restore it then.

**Review by adoroszlai — CHANGES_REQUESTED** [review]

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (DaveTeng0) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java]: sorry it took me a bit time to finish the new unit test! This is restored now.

**Comment by DaveTeng0** [comment]
> Hi @DaveTeng0 , Thanks for the patch. Can you add this check in TestMultiTenantVolume as well which is meant for S3 multi tenant volume scenario.

Yup, I add one test case in TestMultiTenantVolume where ozone should reject non-S3 compliant tenant Id when the default cluster's StrictS3 is true. And for other more different cases, I added in unit test TestOMTenantCreateRequest. Thanks for the review.

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/volume/TestOMVolumeCreateRequest.java (hunk 1 above)
Comment (DaveTeng0) [reply hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/volume/TestOMVolumeCreateRequest.java:213]: Yup, added. Thanks for the review!

**Review by adoroszlai — APPROVED** [review]

**Comment by adoroszlai** [comment]
Thanks @DaveTeng0 for updating the patch.

@ashishkumar50 @devmadhuu please take another look

**Review by ashishkumar50 — APPROVED** [review]
Thanks for updating the patch, LGTM.

**Comment by devmadhuu** [comment]
LGTM +1

**Comment by adoroszlai** [comment]
Thanks @DaveTeng0 for the patch, @ashishkumar50, @devmadhuu for the review.

**Pull request merged**